multigeiger/tube.cpp

```diff
@@ -0,0 +1,116 @@
+  portENTER_CRITICAL_ISR(&mux_GMC_count);
+  static unsigned long isr_count_timestamp_us;
+  static unsigned long isr_count_timestamp_us_prev;
+  static unsigned long isr_count_timestamp_us_prev_used;
+  digitalWrite(PIN_TEST_OUTPUT, HIGH);
+  isr_count_timestamp_us_prev = isr_count_timestamp_us;
+  isr_count_timestamp_us = micros();
+  if ((isr_count_timestamp_us - isr_count_timestamp_us_prev) > GMC_DEAD_TIME) {
+    // the rest is only executed if GMC_DEAD_TIME is exceeded.
+    // Reason: pulses occurring short after another pulse are false pulses generated by the rising edge on the PIN_GMC_COUNT_INPUT.
+    // This happens because we don't have an Schmitt-Trigger on this controller pin.
+    isr_GMC_counts++;                                             // count the pulse
+    isr_count_timestamp = isr_count_timestamp_2send = millis();   // notice (system) time of the pulse
+    isr_GMC_counts_2send++;
+    isr_gotGMCpulse = 1;
+
+    isr_count_time_between = isr_count_timestamp_us - isr_count_timestamp_us_prev_used;  // save for statistics debuging
+    isr_count_timestamp_us_prev_used = isr_count_timestamp_us;
+  }
+  digitalWrite(PIN_TEST_OUTPUT, LOW);
+  portEXIT_CRITICAL_ISR(&mux_GMC_count);
+}
+
+#define DISPLAYREFRESH 10000
+#define MAXCOUNTS 100
+#define HVPULSE_MS 1000
+
+
+// How many HV capacitor charge pulses to generate before giving up.
+// (MAX_CHARGE_PULSES * pulse_duration) should be less than the interval in
+// the main loop (currently 1000ms) where it unconditionally recharges.
+#define MAX_CHARGE_PULSES 333
+
+// How many HV capacitor charge pulses to generate for first charge,
+// for the call in setup(). Not timing critical and capacitor is empty.
+#define MAX_CHARGE_PULSES_INITIAL 3333
+
+int gen_charge_pulses(bool setup) {
+  int max_charge_pulses = setup ? MAX_CHARGE_PULSES_INITIAL : MAX_CHARGE_PULSES;
+  int charge_pulses = 0;
+  isr_GMC_cap_full = 0;
+  do {
+    digitalWrite(PIN_HV_FET_OUTPUT, HIGH);              // turn the HV FET on
+    delayMicroseconds(1500);                            // 5000 usec gives 1,3 times more charge, 500 usec gives 1/20 th of charge
+    digitalWrite(PIN_HV_FET_OUTPUT, LOW);               // turn the HV FET off
+    delayMicroseconds(1000);
+    charge_pulses++;
+  } while ((charge_pulses < max_charge_pulses) && !isr_GMC_cap_full); // either a timeout or a capacitor full interrupt stops this loop
+  time2hvpulse = millis();                              // we just pulsed, so restart timer
+  if ((charge_pulses == max_charge_pulses) && !isr_GMC_cap_full)
+    return -1; // TODO log(CRITICAL, "HV charging failed!");               // pulsed a lot, but still the capacitor is not at desired voltage
+  return charge_pulses;
+}
+
+void setup_tube() {
+  pinMode(PIN_TEST_OUTPUT, OUTPUT);
+  pinMode(PIN_HV_FET_OUTPUT, OUTPUT);
+  pinMode(PIN_HV_CAP_FULL_INPUT, INPUT);
+  pinMode(PIN_GMC_COUNT_INPUT, INPUT);
+
+  digitalWrite(PIN_TEST_OUTPUT, LOW);
+  digitalWrite(PIN_HV_FET_OUTPUT, LOW);
```

Comment:
this line was missing in the original code.

Reply:
Which on of this tree ?

Reply:
Sorry, I don't understand the question.

Reply:
rephrase? oder auf deutsch?

Reply:
Versteh nur Bahnhof.
Du schreibst oben: this line was missing in ...
Und ich frage: which one of this three  (gut, ich hab das e von one vergessen)
Ich weiß nicht, welche Zeile gefehlt hat.
Aber das ist ja eigentlich auch völlig unwichtig.

Reply:
die direkt ueber meinem kommentar.